drivers/scheduler/anthos/anthos.go

```diff
@@ -847,28 +947,129 @@ func getExecPath() (string, error) {
+	if err != nil {
+		return 0, fmt.Errorf("failed get StorageCluster list from namespace [%s], Err: %v", pxNameSpace, err)
+	}
+
+	stc, err := pxOperator.GetStorageCluster(stcList.Items[0].Name, stcList.Items[0].Namespace)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get StorageCluster [%s] from namespace [%s], Err: %v", stcList.Items[0].Name, stcList.Items[0].Namespace, err.Error())
+	}
+	val, ok := stc.Annotations[storagePDBMinAvailable]
+	if !ok {
+		return 0, nil
+	}
+	pdbVal, err := strconv.Atoi(val)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse pdb value. Err: %v", err)
+	}
+	return pdbVal, nil
 }
 
-func (anth *anthos) GetZones() ([]string, error) {
-	// TODO: Add implementation
-	return nil, &errors.ErrNotSupported{
-		Type:      "Function",
-		Operation: "GetZones()",
+// getNodePoolMap return node pool map
+func getNodePoolMap(nodeList []corev1.Node) (map[int][]corev1.Node, error) {
+	nodePoolMap := make(map[int][]corev1.Node)
+	for _, node := range nodeList {
+		poolVal, ok := node.Labels[labelKey]
+		if ok && poolVal != "" {
+			key, err := strconv.Atoi(strings.Split(poolVal, "-")[1])
+			if err != nil {
+				return nodePoolMap, err
+			}
+			poolList, ok := nodePoolMap[key]
+			if !ok {
+				poolList = make([]corev1.Node, 0)
+			}
+			nodePoolMap[key] = append(poolList, node)
+		}
+	}
+	for _, poolNodeList := range nodePoolMap {
+		sort.Slice(poolNodeList, func(i, j int) bool {
+			return poolNodeList[i].CreationTimestamp.Before(&poolNodeList[j].CreationTimestamp)
+		})
 	}
+	return nodePoolMap, nil
 }
 
-func (anth *anthos) GetASGClusterSize() (int64, error) {
-	// TODO: Add implementation
-	return 0, &errors.ErrNotSupported{
-		Type:      "Function",
-		Operation: "GetASGClusterSize()",
+// getSequentialUpgradeTime calculate upgrade times for sequential upgrade
+func getSequentialUpgradeTime(upgradeStartTime time.Time, sortedNodes []corev1.Node) error {
+	return printUpgradeTimeExceededErrorMessage(printNodeUpgradeTime(upgradeStartTime, sortedNodes))
+}
+
+// getParallelUpgradeTime calculate upgrade times for sequential upgrade
+func getParallelUpgradeTime(upgradeStartTime time.Time, sortedNodes []corev1.Node, maxNode int) error {
+	errorMessages := make([]string, 0)
+	var newUpgradeStartTime time.Time
+	nodePoolMap, err := getNodePoolMap(sortedNodes)
+	if err != nil {
+		return fmt.Errorf("fail to retrieve node pool map. Err: %v", err)
+	}
+	log.Debugf("Retrieved node pool map: %v", nodePoolMap)
+	for pool := 0; pool < len(nodePoolMap); pool++ {
+		startTime := upgradeStartTime
+		errorMessages = append(errorMessages, printNodeUpgradeTime(startTime, nodePoolMap[pool])...)
+		// When number of pools is more than the number of nodes can be upgraded simultaneously.
+		// In that case upgrade will not start simultaneously in all node pools.
+		// Finding the new upgrade start time for node pools where upgrade will start later
+		if maxNode < len(nodePoolMap) {
```

Comment:
This logic needs to be modified for more number of node pools. Use the following:
```
nodePoolMap[i%maxNode] = ....
```

Reply:
Incorporated